sync2/fptree/fptree.go

```diff
@@ -0,0 +1,1295 @@
+	// errEasySplitFailed if easySplit is enabled, so we can ignore them here
+	middle := make(rangesync.KeyBytes, ft.keyLen)
+	ac.lastPrefix0.idAfter(middle)
+	ft.log("easySplit: lastPrefix0 %s middle %s", ac.lastPrefix0, middle)
+	items := ft.From(x, 1)
+	part0 := FPResult{
+		FP:    ac.fp0,
+		Count: ac.count0,
+		IType: ac.itype,
+		Items: items,
+		// Next is only used during splitting itself, and thus not included
+	}
+	items = ft.startFromPrefix(&ac, *ac.lastPrefix0)
+	part1 := FPResult{
+		FP:    ac.fp,
+		Count: ac.count,
+		IType: ac.itype,
+		Items: items,
+		// Next is only used during splitting itself, and thus not included
+	}
+	return SplitResult{
+		Part0:  part0,
+		Part1:  part1,
+		Middle: middle,
+	}, nil
+}
+
+// Split splits an interval in two parts.
+func (ft *FPTree) Split(x, y rangesync.KeyBytes, limit int) (sr SplitResult, err error) {
+	ft.np.lockRead()
+	defer ft.np.unlockRead()
+	sr, err = ft.easySplit(x, y, limit)
+	if err == nil {
+		return sr, nil
+	}
+	if err != errEasySplitFailed {
+		return SplitResult{}, err
+	}
+
+	fpr0, err := ft.fingerprintInterval(x, y, limit)
+	if err != nil {
+		return SplitResult{}, err
+	}
+
+	if fpr0.Count == 0 {
+		return SplitResult{}, errors.New("can't split empty range")
+	}
+
+	fpr1, err := ft.fingerprintInterval(fpr0.Next, y, -1)
+	if err != nil {
+		return SplitResult{}, err
+	}
+
+	if fpr1.Count == 0 {
+		return SplitResult{}, errors.New("split produced empty 2nd range")
+	}
+
+	return SplitResult{
+		Part0:  fpr0,
+		Part1:  fpr1,
+		Middle: fpr0.Next,
+	}, nil
+}
+
+// dumpNode prints the node structure to the writer.
+func (ft *FPTree) dumpNode(w io.Writer, idx nodeIndex, indent, dir string) {
+	if idx == noIndex {
+		return
+	}
+
+	count, fp, leaf := ft.np.info(idx)
+	countStr := strconv.Itoa(int(count))
+	if leaf {
+		countStr = "LEAF:" + countStr
+	}
+	var valStr string
+	if v := ft.np.value(idx); v != nil {
+		valStr = fmt.Sprintf(" <val:%s>", shortened(v))
+	}
+	fmt.Fprintf(w, "%s%sidx=%d %s %s [%d]%s\n", indent, dir, idx, fp, countStr, ft.np.refCount(idx), valStr)
+	if !leaf {
+		indent += "  "
+		ft.dumpNode(w, ft.np.left(idx), indent, "l: ")
+		ft.dumpNode(w, ft.np.right(idx), indent, "r: ")
+	}
+}
+
+// Dump prints the tree structure to the writer.
+func (ft *FPTree) Dump(w io.Writer) {
+	ft.np.lockRead()
+	defer ft.np.unlockRead()
+	if ft.root == noIndex {
+		fmt.Fprintln(w, "empty tree")
+	} else {
+		ft.dumpNode(w, ft.root, "", "")
+	}
+}
+
+// DumpToString returns the tree structure as a string.
+func (ft *FPTree) DumpToString() string {
+	var sb strings.Builder
+	ft.Dump(&sb)
+	return sb.String()
+}
+
+// Count returns the number of items in the tree.
+func (ft *FPTree) Count() int {
+	ft.np.lockRead()
+	defer ft.np.unlockRead()
+	if ft.root == noIndex {
+		return 0
+	}
+	return int(ft.np.count(ft.root))
+}
+
+// EnableTrace enables or disables tracing for the tree.
+func (ft *FPTree) EnableTrace(enable bool) {
+	ft.traceEnabled = enable
+}
```

Comment:
i think that reviewing this file while understanding everything which is happening here will probably take me a week. i will therefore "trust you on this"

Reply:
Not insisting on reading this right now, but for future reference, the docs on how FPTree works are in the `dev-docs` folder.
https://github.com/spacemeshos/go-spacemesh/blob/44d888f22e96d0f1b0a59480af19548b23ba38de/dev-docs/sync2-set-reconciliation.md#fptree-data-structure